### PR TITLE
AirfRANS: Baseline replication with seed=789 at 4L/256d golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
AirfRANS exhibits stochastic phase transitions — seed variance is high. Seed 789 was the best performer in multi-seed characterization at 3L/192d. Testing whether seed=789 at 4L/256d + golden config + full 180-min budget beats the current baseline (which was epoch-capped at 50 epochs). This also serves as a replication of the baseline with proper SENPAI_MAX_EPOCHS=9999 and SENPAI_TIMEOUT_MINUTES=180.

## Instructions
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --seed 789 \
  --agent nezuko \
  --wandb-name "nezuko/airfrans-4L256d-seed789-golden" \
  --wandb-group "airfrans-seed-sweep-golden"
```

Report val_primary/surface_mse, best epoch, total epochs, and time per epoch.

## Baseline
Current best: **0.007264** at epoch 40 (4L/256d, golden config, 50 epochs in 61 min — epoch capped!)
Note: Baseline hit SENPAI_MAX_EPOCHS=50 cap. With 9999 cap and 180-min, we should get ~150 epochs.
External target: 0.0043

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Key Metrics

| Metric | Value | Baseline (PR) | External Target | vs Baseline | vs External |
|--------|-------|---------------|-----------------|-------------|-------------|
| val_primary/surface_mse | **0.003392** | 0.007264 | 0.0043 | **-53.3%** | **-21.1%** |
| Best epoch | 161 | 40 | — | — | — |
| Total epochs | 223 | 50 (capped) | — | — | — |
| Time per epoch | ~48.6s | — | — | — | — |

### Component Metrics at Best Epoch (161)

| Component | Value |
|-----------|-------|
| surface_mse_p (pressure) | 0.01348 |
| surface_mse_Ux | 5.60e-05 |
| surface_mse_Uy | 5.35e-06 |
| surface_mse_nut | 2.61e-05 |
| volume_mse | 0.01230 |

### Training Dynamics

- **Phase transitions**: Epoch 18 (0.14→0.06), epoch 39 (0.037→0.016), epoch 50 (0.033→0.009)
- **Trough progression** (cosine LR cycling, T_max=5): ep50→0.00928, ep75→0.00645, ep90→0.00563, ep110→0.00421, ep129→0.00348, ep161→0.00339
- **Divergence**: Training diverged at epoch 177 (grad_norm spiked from ~3 to 18.76), became terminal at ep189 (grad_norm=5.4B), NaN from ep192 onward
- **Grad norm trend** (pre-divergence): Steadily decreased from ~24 (ep20) → ~3 (ep170)
- **Loss trend** (pre-divergence): Decreased from ~0.42 (ep20) → ~0.03 (ep170)

### Test Metrics

Test metrics are **NaN** because test evaluation runs on the final model state, and the model was in a NaN-diverged state by training end (ep223). There is no best-checkpoint reloading in train.py. This is a known limitation that could be addressed in a follow-up.

### Peak Memory

- GPU: ~8.2 GB VRAM on a single GPU (CUDA_VISIBLE_DEVICES=0)

### W&B

- Run ID: `ccpv82op`
- Run URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ccpv82op

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 7e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --seed 789 \
  --agent nezuko \
  --wandb-name "nezuko/airfrans-4L256d-seed789-golden" \
  --wandb-group "airfrans-seed-sweep-golden"
```

### What Happened

**The experiment succeeded.** Seed 789 with the golden config (4L/256d, gc=1.0) and extended 180-min budget achieved val_primary/surface_mse of **0.003392**, beating:
- The PR baseline (0.007264) by 53.3%
- The external target (0.0043) by 21.1%

The extended training budget was critical — the previous 50-epoch cap would have yielded only 0.009279 (ep50). The best result came at epoch 161, well beyond the old cap. The cosine LR cycling with T_max=5 produced progressively deeper troughs over time, confirming that longer training with this schedule continues to find better minima.

However, training diverged at epoch 177 (gradients exploded), suggesting the 7e-4 learning rate may be at the stability boundary for very long runs. The divergence pattern — periodic grad_norm spikes every ~20 epochs (at ep100, 120, 140, 160, 180) growing in amplitude — points to accumulating instability in the cosine cycling.

### Suggested Follow-ups

1. **Best-checkpoint reloading for test evaluation**: train.py should save the best checkpoint and reload it for final test evaluation. Currently, test metrics are lost if training diverges.
2. **LR warmdown or reduction**: A warmdown schedule after ~150 epochs could prevent the late divergence while preserving the gains.
3. **Gradient clipping tightening**: Reducing grad_clip from 1.0 to 0.5 (as in the gc=0.5 experiments) may stabilize longer runs.
4. **Seed 789 with gc=0.5**: Combining this seed with the proven gc=0.5 config could yield even better results with more stability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2808 because ADVISOR: Closing — your 4L/256d seed=789 replication scored 0.003392, well above the 0.001479 baseline. This confirms the 4L/256d architecture is consistently worse than 3L/256d regardless of seed. Useful reproducibility check, nezuko..
